### PR TITLE
Feature/evm reader handle epochs

### LINKED ISSUE #545: [EVM-Reader] Add Epoch Handling
## 📚 Context

Node needs to keep track of the Epoch lifecycle correctly. EVM Reader is the main component to retrieve information from the blockchain and the component responsible for reading inputs, and so, making the node move forward

## ✔️ Solution

EVM Reader needs to materialize Epochs entities and keep track of inputs in each epoch

### PR DESCRIPTION
**Pull request by fmoura**
closes #545

### CODE REVIEW

File: internal/evmreader/evmreader.go [hunk 1]
```diff
@@ -115,6 +139,9 @@ func (r *EvmReader) Run(
 	ready chan<- struct{},
 ) error {
 
+	// Initialize epochLength cache
+	r.epochLengthCache = make(map[Address]uint64)
```
[the remark this answers is not in the document]
Comment (fmoura) [reply internal/evmreader/evmreader.go:143]: Epoch's lengths should not change, so the cache is built once per service run

File: internal/evmreader/evmreader.go [hunk 2]
```diff
@@ -115,6 +139,9 @@ func (r *EvmReader) Run(
...
+
 	for {
 		watchForNewInputsError := r.watchForNewBlocks(ctx, ready)
```
[the remark this answers is not in the document]
Comment (fmoura) [reply internal/evmreader/evmreader.go]: It was intended to be the same name of the called method. After a few renames, it was not quite right.

Keeping it just `err` as usual is the best way. Thanks

**Review by vfusco — CHANGES_REQUESTED** [review]

**Review by vfusco — DISMISSED** [review]

File: internal/evmreader/evmreader.go [hunk 3]
```diff
@@ -256,105 +271,378 @@ func (r *EvmReader) fetchMostRecentHeader(
...
+
+func appToAddresses(apps []Application) []Address {
+	var addresses []Address
+	for _, app := range apps {
+		addresses = append(addresses, app.ContractAddress)
+	}
+	return addresses
+}
+
+// Insert Sort for Inputs into a Input slice
```
Comment (torives) [inline internal/evmreader/evmreader.go:638]:
```suggestion
// insertSorted inserts the received input in the slice at the position defined by its index property.
```

File: internal/evmreader/evmreader.go [hunk 4]
```diff
@@ -256,105 +271,378 @@ func (r *EvmReader) fetchMostRecentHeader(
...
+
+// Util functions
+
+// Calculates the epoch index given the input block number
+func calculateEpochIndex(epochLength uint64, blockNumber uint64) uint64 {
+	return blockNumber / epochLength
+}
+
+// Closes the epochs in the given interval
+// If includeLastEpoch param tells if the last epoch should be closed or not
```
Comment (torives) [inline internal/evmreader/evmreader.go:607]:
```suggestion
// closeEpochs closes all epochs within the given index interval, except for the last.
// If `includeLastEpoch` is true, the last one is included.
```

File: internal/evmreader/retrypolicy/contratfactory.go [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+		maxRetries: maxRetries,
+		maxDelay:   maxDelay,
+	}
+}
+
+func (f *EvmReaderContractFactory) NewApplication(
+	address common.Address,
+) (evmreader.ApplicationContract, error) {
+
+	// Building a contract do not fail due to network errors.
```
Comment (torives) [inline internal/evmreader/retrypolicy/contratfactory.go:40]:
```suggestion
	// Building a contract does not fail due to network errors.
```

File: internal/evmreader/retrypolicy/contratfactory.go [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+
+	return NewApplicationWithRetryPolicy(applicationContract, f.maxRetries, f.maxDelay), nil
+
+}
+
+func (f *EvmReaderContractFactory) NewIConsensus(
+	address common.Address,
+) (evmreader.ConsensusContract, error) {
+
+	// Building a contract do not fail due to network errors.
```
Comment (torives) [inline internal/evmreader/retrypolicy/contratfactory.go:55]:
```suggestion
	// Building a contract does not fail due to network errors.
```

**Review by torives — COMMENTED** [review]
Some minor things to fix before the merge:

File: internal/retry/retry.go [hunk 7]
```diff
@@ -26,6 +26,12 @@ func CallFunctionWithRetryPolicy[
 	var lastValue R
 
 	for i := uint64(0); i <= maxRetries; i++ {
+
+		if i != 0 {
+			slog.Info("Retry Policy: Retrying...", "delay", maxDelay)
```
[the remark this answers is not in the document]
Comment (fmoura) [reply internal/retry/retry.go:31]: We want to know that. It was originally Debug, but became an Info on a previous PR

File: internal/evmreader/evmreader.go [hunk 8]
```diff
@@ -256,105 +271,378 @@ func (r *EvmReader) fetchMostRecentHeader(
...
+				err)
 		}
+		r.epochLengthCache[app.ContractAddress] = epochLength
+		slog.Info("Got epoch length from IConsensus",
+			"app", app.ContractAddress,
+			"epoch length", epochLength)
+	} else {
+		slog.Info("Got epoch length from cache",
+			"app", app.ContractAddress,
+			"epoch length", epochLength)
```
[the remark this answers is not in the document]
Comment (fmoura) [reply internal/evmreader/evmreader.go:458]: Almost the same as above. We want more information when running with "Info" log level. If it really becomes too much we can downgrade it to "Debug" .

The "cached" one makes sense to downgrade.

**Review by torives — APPROVED** [review]

**Review by vfusco — APPROVED** [review]

**Pull request merged**